Script to create appvm to run haveno on qubes

BrandyJSon opened the pull request:
This script allows a user to easily create a dedicated qube to run haveno. By default it will pull the official repository but can also be used to install unofficial repositories to access mainnet trading. A user can specify to either use a published deb.zip release (check hashes automatically, and signatures if key is supplied) or build from source.  The script also does some basic hardening on debian-12-minimal template to protect the user from being de anonymized. 

Please let me know if more documentation or detail in the README is necessary.

woodser commented on scripts/install_qubes/haveno-qubes-install.sh:
```diff
@@ -0,0 +1,228 @@
+	log "Compilation successful, creating a script to run compiled binary securly"
+	$TPL_ROOT "echo -e '#\x21/bin/bash\nsource /root/.sdkman/bin/sdkman-init.sh\n/root/$CODE_DIR/haveno-desktop --torControlPort=9051 --useTorForXmr=OFF --torControlHost=$SYS_WHONIX_IP' > /usr/local/sbin/Haveno && chmod +x /usr/local/sbin/Haveno && chmod u+s /usr/local/sbin/Haveno"
+	#Fix icon permissions
+	$TPL_ROOT "cp '/root/haveno-reto/desktop/package/linux/haveno.png' /opt/haveno.png && chmod 644 /opt/haveno.png"
```
Please remove haveno-reto.

BrandyJSon commented:
Sorry about that, removed haveno-reto in the most recent commit.

woodser commented:
@boldsuck had some good feedback he posted in the dev chat:

```
<b​oldsuck> I posted in the Reto Chat a few days ago that someone should test it. Tinted again today
<b​oldsuck> idk if brandyjson's script need --torControlUseSafeCookieAuth bc torControl port always needs auth.
<b​oldsuck> He should perhaps take a look at my DirectBindTor. Most people don't know that it exists and rework bisq scripts.
<b​oldsuck> In line 291 he has:
<b​oldsuck> qvm-run --pass-io -u root -- whonix-gateway-17 "echo -e 'Sandbox 1\nConnectionPadding 1\n' > /usr/local/etc/torrc.d/50_user.conf"
<b​oldsuck> why not
<b​oldsuck> qvm-run --pass-io -u root -- whonix-gateway-17 "echo -e 'Sandbox 1\nConnectionPadding 1\nHiddenServiceDir /var/lib/tor/haveno_service/1\nHiddenServicePort 9999 "$(qvm-prefs sys-whonix ip)":9999' > /usr/local/etc/torrc.d/50_user.conf"
<b​oldsuck> And leave out all those with Tor ControlPort and onion grater. Just ust default SocksPort:9050
<b​oldsuck> TorControl Port is even more powerful than monerod --rpc-bind-port
<b​oldsuck> I have 'ControlPort 0' on Tor relays. How secure is network traffic between Qubes VM on unsecure OS? idk
<b​oldsuck> Whonix-gateway was able to use clipboard sharing with Host OS per default
```

BrandyJSon commented:
Just updated the script to use boldsuck's suggested method. No more control port access needed. 

Additionally I added another script to make it easy for a user to generate a new hidden service address for haveno to utilize.

woodser commented:
Feedback from syringes in the haveno chat channel:

> I tested the script out on a fresh Qubes installation. I successfully installed and ran retoswap but there were a few bugs.
>
> Fails to restart sys-whonix if VMs are connected. Standard Qubes behavior but is it possible to mimic the restart functions in Qubes Update?
>
> Signature failed multiple times when entered 'https://retoswap.com/reto_public.asc' at the signature prompt but was able to verify the .sig when manually downloaded in a dispvm
>
> Replacing the segment at line 126 with this fixed the issue
> ```
> if [[ $from_source -eq 1 ]]; then
> 	log "Downloading haveno release version $version"
> 	$TPL_ROOT "curl -L  --remote-name-all $HAVENO_REPO/releases/download/$version/{$TARGET_DEB,$TARGET_DEB.sig,$version-hashes.txt}"
> 	read -p "Enter url to verify signatures or anything else to skip:" key
> 	if [[ $key =~ $regex ]]; then
> 		$TPL_ROOT "apt-get install --no-install-recommends gnupg2 -y"
> 		$TPL_ROOT "gpg2 --fetch-keys $key"
> 		verify_output=$($TPL_ROOT "gpg2 --verify $TARGET_DEB.sig 2>&1")
> 	if [[ $verify_output =~ "Good signature" ]]; then
> 		log "Signature valid, continuing"
> 	else
> 		log "Signature invalid, exiting"
> 		exit 1
> 	fi
> ```
> 
> Hardening did not work with tirdad. Even after modifying for the correct repositories and public key an error code 1 occured and stopped the script.
> Perhaps comment out tirdad until then?
> Also
>
> on line 219 `if [ $unhardened ]; then`
> should be
> `if [ $unhardened -eq 0 ]; then`
> Also eventually if anyone would like to make a guide for manual installation. I believe a lot more people would feel comfortable with that rather than running a script in dom0
```

boldsuck commented:
The script looks good, unfortunately I can't test it at the moment.
I've already written it in the Reto SimpleX chat and will ask for testers in the Whonix forum.

@BrandyJSon

**Not for this script**, just as a hint if useful elsewhere, bc line 104:
`read -p "Warning by default 50_user.conf on sys-whonix will be overwritten (baseline is empty`

There is an `%include` function for torrc, e.g:
`%include /usr/local/etc/torrc.d/foobar`
or from Haveno options, for example:
`/opt/haveno/bin/Haveno --torrcOptions=%include /home/user/.local/share/Haveno/torrc.local`

The latter is useful for the default Haveno installation (using netlayer/jtorctl) if users want to configure their own tor options, e.g: logging
If you don't use `DirectBindTor,` Haveno rewrites the torrc in the appDataDir on every restart and therefore direct editing is useless.

boldsuck commented:
> I tested the script out on a fresh Qubes installation. I successfully installed and ran retoswap but there were a few bugs.

It's nice to hear that it works on Qubes via SocksPort.

> Signature failed multiple times when entered 'https://retoswap.com/reto_public.asc' at the signature prompt but was able to verify the .sig when manually downloaded in a dispvm

RetoSwap is on Github. This is probably more of an issue because Github is Tor hostile. I'll ask the arbs if they can upload the key to keys.openpgp.org. Then you can use:
`keyserver hkp://zkaan2xfbuxia2wpf7ofnkbz6r5zdbbvxbunvp5g2iebopbfc4iqmbad.onion`

BrandyJSon commented:
Everything mentioned above should be fixed, signature checks & tirdad installation.

I believe this should be ready for merge but feel free to let me know if anything else needs changes.

boldsuck commented:
I format the guide until I like it. For Haveno Docu I will of course change the reto installation instructions.
[I have pointed out this PR](https://boldsuck.github.io/haveno-docs/haveno-ui/DirectBindTor/#qubes-os)

[Whonix Wiki](https://www.whonix.org/wiki/Onion_Services#Why_use_Whonix_for_Hosting_Onion_Services) also points out that [ZeroNet, OnionShare and Bisq talking to Tor control port ](https://www.whonix.org/wiki/Onion_Services#cite_note-3) is not 'optimal'.
@BrandyJSon I like your script. 

Maybe someone will have time to extend `DirectBindTor` at some point and adds Unix socket support to TCP socket 'SocksPort:9050'. But I have no idea if this is possible with a Java app.

xyhhx commented:
discovered this as someone shared it in the qubes os room.

i was curious about a few things:

for @BrandyJSon:

- why did you choose to offer bash scripts instead of salt states?
- to that end, why not offer rpm packages so these can be installed into a qube using ootb tools?
- nitpick, but why did you choose to format lines [like this](https://github.com/haveno-dex/haveno/commit/30956ac03fa8dd60aadbd447501ae529f181e1b9#diff-0d597810349fba8e7c47615eb5d27e5a4775e170843add990287272ccf0d56e4R171)? they're hard to read
- why not use unix sockets for the `socat` tunnels?
- why not use more of kicksecure's offerings?
- you should use `qubesdb-read /name` instead of `hostname`, as it's less likely to be tampered with in a disaster scenario
- i _want_ to say "why don't you verify the commits of the git repos you're cloning?", but it looks like haveno doesn't even bother to sign their commits lol :clown_face: 
- it might be a better idea to move the `git clone` and other download commands to a disposable vm and only copy them to the template once they're verified etc
- the template isn't very hardened in my opinion. the only steps you took was installing tirdad and removing a few packages at the end of the installer from the template (including gpg??)
- qubes split-gpg is not configured, which would reduce the chances of compromising your gpg toolchain in the haveno template/appvm

anyway those were just some notes i had after perusing the script

@boldsuck i was wondering about you choice in distro too, if you don't mind

- why are you specifically interested in debian? it is far more out of date than fedora, and fedora has selinux which is more robust than apparmor

xyhhx commented:
fair enough!

woodser commented:
Closing as another approach was preferred based on community feedback: https://github.com/haveno-dex/haveno/pull/1628

Please feel free to provide additional feedback on those instructions, by opening an issue or chatting in Matrix.
